server/src/main/java/io/druid/segment/indexing/Transformer.java

```diff
@@ -0,0 +1,232 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.segment.indexing;
+
+import com.google.common.base.Strings;
+import io.druid.data.input.InputRow;
+import io.druid.data.input.Row;
+import io.druid.java.util.common.DateTimes;
+import io.druid.math.expr.Expr;
+import io.druid.query.filter.ValueMatcher;
+import io.druid.query.groupby.RowBasedColumnSelectorFactory;
+import io.druid.segment.column.Column;
+import org.joda.time.DateTime;
+
+import javax.annotation.Nullable;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ *
+ */
+public class Transformer
+{
+  private final Map<String, Expr> transforms = new HashMap<>();
+  private final ThreadLocal<Row> rowSupplierForValueMatcher = new ThreadLocal<>();
+  private final ValueMatcher valueMatcher;
+
+  Transformer(final TransformSpec transformSpec)
+  {
+    for (final Transform transform : transformSpec.getTransforms()) {
+      transforms.put(transform.getName(), transform.toExpr());
+    }
+
+    if (transformSpec.getFilter() != null) {
+      valueMatcher = transformSpec.getFilter().toFilter()
+                                  .makeMatcher(
+                                      RowBasedColumnSelectorFactory.create(
+                                          rowSupplierForValueMatcher,
+                                          null
+                                      )
+                                  );
+    } else {
+      valueMatcher = null;
+    }
+  }
+
+  /**
+   * Transforms an input row, or returns null if the row should be filtered out.
+   *
+   * @param row the input row
+   */
+  @Nullable
+  public InputRow transform(@Nullable final InputRow row)
+  {
+    if (row == null) {
+      return null;
+    }
+
+    final InputRow transformedRow;
+
+    if (transforms.isEmpty()) {
+      transformedRow = row;
+    } else {
+      transformedRow = new TransformedInputRow(row, transforms);
+    }
+
+    if (valueMatcher != null) {
+      rowSupplierForValueMatcher.set(transformedRow);
+      if (!valueMatcher.matches()) {
+        return null;
```

Comment:
most indexers don't have the ability to filter null rows but rather fail with NPE , does this filtering work with either hadoop indexing job or kafka indexing service ?

Reply:
@himanshug that's a good point, I tested it with IndexTask but not every form. IndexTask works since it has this code:

```
          // The null inputRow means the caller must skip this row.
          if (inputRow == null) {
            continue;
          }
```

I'll add some tests for kafka, hadoop, and realtime tasks as well.